File: 2019/04/solution.py
#!/usr/bin/python3

import itertools

def parse_range(input):
    return tuple(int(i) for i in input.strip().split('-'))

def digits(integer):
    return [int(digit) for digit in str(integer)]

def pairs(xs):
    yield from zip(xs, xs[1:])

def are_increasing(xs):
    return all(a <= b for a, b in pairs(xs))

def has_doubles(xs):
    return any(a == b for a, b in pairs(xs))

def group_lengths(xs):
    yield from (len(list(group)) for key, group in itertools.groupby(xs))

def task1(input):
    start, end = parse_range(input)
    return sum(
        1 for i in range(start, end)
        if are_increasing(digits(i))
        and has_doubles(digits(i))
    )

def task2(input):
    start, end = parse_range(input)
    return sum(
        1 for i in range(start, end)
        if are_increasing(digits(i))
        and 2 in group_lengths(digits(i))
    )

input = open('input.txt', 'r').read()
print(task1(input))
print(task2(input))

File: 2019/04/solution.hs
#!/usr/bin/runghc

import Data.List

main :: IO ()
main = do
    input <- readFile "input.txt"
    putStrLn $ task1 input
    putStrLn $ task2 input

task1 :: String -> String
task1 =
    show
    . length
    . filter hasDoubles
    . filter isIncreasing
    . map digits
    . parseRange

task2 :: String -> String
task2 =
    show
    . length
    . filter (any (== 2))
    . map groupLengths
    . filter isIncreasing
    . map digits
    . parseRange

parseRange :: String -> [Int]
parseRange range =
    let [start, end] = map read $ splitOn '-' range
    in [start .. end]

splitOn :: Char -> String -> [String]
splitOn delim str =
    case str of
        "" -> []
        s' -> let (x, s'') = break (== delim) s'
              in x : splitOn delim (drop 1 s'')

digits :: Int -> [Int]
digits n
    | n < 10 = [n]
    | otherwise = digits (n `div` 10) ++ [n `mod` 10]

pairMap :: (a -> a -> b) -> [a] -> [b]
pairMap f xs =
    zipWith f xs $ tail xs

isIncreasing :: Ord a => [a] -> Bool
isIncreasing xs =
    all id $ pairMap (<=) xs

hasDoubles :: Eq a => [a] -> Bool
hasDoubles xs =
    any id $ pairMap (==) xs

groupLengths :: Eq a => [a] -> [Int]
groupLengths xs =
    map length $ group xs
